### PR TITLE
[WIP] Warning when passing padded input ids but no attention mask

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

One of the most common mistake users make in Transformers IMO is that `input_ids` are padded, but no `attention_mask` is provided (we see many examples of this). As discussed multiple times, we **don't** want to infer the `attention_mask` automatically as this creates a lot of unmaintainable, "not-possible-to-deal-with" complexity.

A while ago, we discussed to throw a warning in this case, making sure it's done only once to not spam the user when calling the model multiple times. I'm not sure we found a good conclusion, but IMO it's important that we warn the user as too users (IMO) think the attention_mask is inferred from the padding tokens. This PR is tries to solve this and shows how it'd be implemented for just BERT. We would have to implement it for all other models then as well. Would very much like to hear your opinion here @sgugger @LysandreJik @patil-suraj  . Note that this PR will touch a lot of important functions / files, so it'd be very important to make the warning as clear as possible. 
I do however have a strong conviction that we should display such a warning.

No the warning function can display the following warning messages for a toy BERT example of passing just three input ids.

Possible warning messages:

1. Pad token present, no attention mask, eos, bos, sep all different from pad (that's **VERY** likely an error IMO):

**Displayed warning:**
```
The input IDs tensor([[0, 1, 1]]) contains the `pad_token_id` 0, but NO `attention_mask` is passed.
Padding the input IDs without passing an `attention_mask` leads to unexpected, possibly incorrect outputs.
```

2. Pad token present, no attention mask, eos or bos or sep same as pad:

**Displayed warning:**
```
The input IDs tensor([[0, 1, 1]]) contains the `pad_token_id` 0, but NO `attention_mask` is passed.
We strongly recommend passing an `attention_mask` to avoid possibly incorrectly computing the attention weights. 
You can ignore this warning, if your `pad_token_id` 0 is identical to your `sep_token_id` 0 AND your input is NOT padded.
```

3. Pad token present, no attention mask, two or more of eos, bos, sep identical to pad (don't think this exists actually):

**Displayed warning:**
```
The input IDs tensor([[0, 1, 1]]) contains the `pad_token_id` 0, but NO `attention_mask` is passed.
We strongly recommend passing an `attention_mask` to avoid possibly incorrectly computing the attention weights. 
You can ignore this warning, if your `pad_token_id` 0 is identical to your `bos_token_id` 0 AND your input is NOT padded.
We strongly recommend passing an `attention_mask` to avoid possibly incorrectly computing the attention weights. 
You can ignore this warning, if your `pad_token_id` 0 is identical to your `sep_token_id` 0 AND your input is NOT padded.
```

4. Otherwise no warning.

Also note that the warning only appears at the first forward call.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Relevant issues:
https://github.com/huggingface/transformers/issues/4083
https://github.com/huggingface/transformers/issues/278
https://github.com/huggingface/transformers/issues/16136

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1006,6 +1006,72 @@ def get_input_embeddings(self) -> nn.Module:
...
+            )
+
+            if is_pad_token_equal_to_bos_token:
+                warn_string += (
+                    "\nWe strongly recommend passing an `attention_mask` to avoid possibly incorrectly computing the"
+                    " attention weights. \nYou can ignore this warning, if your `pad_token_id`"
+                    f" {self.config.pad_token_id} is identical to your `bos_token_id` {self.config.bos_token_id} AND"
+                    " your input is NOT padded."
+                )
+            if is_pad_token_equal_to_eos_token:
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:1052]: maybe use `elif` here and below ?

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1006,6 +1006,72 @@ def get_input_embeddings(self) -> nn.Module:
         else:
             raise NotImplementedError
 
+    def warn_if_pad_token_in_input_ids_no_attention_mask(self, input_ids, attention_mask):
+        if attention_mask is not None:
+            # if the attention_mask is defined it's all good
+            return
+
+        if not hasattr(self, "warnings_issued"):
+            self.warnings_issued = {}
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:1015]: Do we try to avoid adding new instance attribute in `__init__`?

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py:1052]: Could be the case that pad_token == eos_token == bos_token -> would like to append the string then

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -1006,6 +1006,72 @@ def get_input_embeddings(self) -> nn.Module:
...
+                    " your input is NOT padded."
+                )
+            if is_pad_token_equal_to_sep_token:
+                warn_string += (
+                    "\nWe strongly recommend passing an `attention_mask` to avoid possibly incorrectly computing the"
+                    " attention weights. \nYou can ignore this warning, if your `pad_token_id`"
+                    f" {self.config.pad_token_id} is identical to your `sep_token_id` {self.config.sep_token_id} AND"
+                    " your input is NOT padded."
+                )
+            if not (is_pad_token_equal_to_bos_token or is_pad_token_equal_to_eos_token):
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:1066]: lack `is_pad_token_equal_to_sep_token` here, I guess?

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -1006,6 +1006,72 @@ def get_input_embeddings(self) -> nn.Module:
...
+                "but NO `attention_mask` is passed."
+            )
+
+            if is_pad_token_equal_to_bos_token:
+                warn_string += (
+                    "\nWe strongly recommend passing an `attention_mask` to avoid possibly incorrectly computing the"
+                    " attention weights. \nYou can ignore this warning, if your `pad_token_id`"
+                    f" {self.config.pad_token_id} is identical to your `bos_token_id` {self.config.bos_token_id} AND"
+                    " your input is NOT padded."
+                )
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1051]: Maybe here let's check if the pad token ID is only used once per input at the beginning before throwing the warning?

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -1006,6 +1006,72 @@ def get_input_embeddings(self) -> nn.Module:
         else:
             raise NotImplementedError
 
+    def warn_if_pad_token_in_input_ids_no_attention_mask(self, input_ids, attention_mask):
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1009]: Let's not push descriptive names too far ;-)  I think `padding_attention_mask_warning` is more than enough!

**Review by sgugger — APPROVED** [review]
Thanks for adding those warnings (mainly the last one). I'd like to add more tests for the warnings when the pad token is the same as the eos/bos/sep tokens to avoid scaring a user for nothing (users are always scared of warnings) and it shouldn't hurt performance since they would only be run once.

As for @LysandreJik comment, I must admit I don't understand what you're suggesting Lysandre, since we only have those pad/eos/bos/sep token IDs from the config of the mdoel inside the forward.

**Comment by LysandreJik** [comment]
Sounds good, I'm likely worrying for nothing then. Good for me like this, very easy to add kwargs afterwards anyway!

**Comment by LysandreJik** [comment]
I think this would be an impactful addition! @ydshieh, would you be interested in continuing this PR?

**Comment by ydshieh** [comment]
> I think this would be an impactful addition! @ydshieh, would you be interested in continuing this PR?

Sure. I will take a look and see if there is anything blocking.

**Comment by ydshieh** [comment]
You can search `elif input_ids is not None:` that is in the base model classes like `BertModel` (already done by @patrickvonplaten), `GPT2Model` etc.

You don't need to replace all of them - it would be super nice already for a few of the most used modes 🚀  Thank you!
